File: Source/Aladdin.ISO7816/DataObjectFile.cs
using System;

namespace Aladdin.ISO7816
{
    ///////////////////////////////////////////////////////////////////////////////
    // Элементарный файл объектов
    ///////////////////////////////////////////////////////////////////////////////
    public class DataObjectFile : ElementaryFile
    {
        // конструктор
        internal DataObjectFile(DedicatedFile parent, ushort id) : base(parent, id) {}

        // конструктор
        internal DataObjectFile(DedicatedFile parent, byte shortID) : base(parent, shortID) {}

        // конструктор
        internal DataObjectFile(DedicatedFile parent, ushort? id, byte? shortID, 

            // сохранить переданные параметры
            BER.FileControlInformation info) : base(parent, id, shortID, info) {}

        // структура файла
        public override FileStructure FileStructure { get 
        {
            // получить дескриптор файла
            DataObject[] objs = Info[Tag.Context(0x02, ASN1.PC.Primitive)]; 
            
            // проверить наличие дескриптора
            if (objs.Length == 0) return ISO7816.FileStructure.DataObject; 

            // получить содержимое
            byte[] content = objs[0].Content; 
            
            // проверить размер содержимого
            if (content.Length < 1 || (content[0] & 0x80) != 0)
            {
                // указать значение по умолчанию
                return ISO7816.FileStructure.DataObject; 
            }
            // в зависимости установленных битов
            if (((content[0] >> 3) & 0x7) == 0x7)
            {
                // в зависимости установленных битов
                switch (content[0] & 0x7)
                {
                case 0x1: return ISO7816.FileStructure.DataObjectBERTLV;
                case 0x2: return ISO7816.FileStructure.DataObjectSimpleTLV;
                }
            }
            // структура файла неизвестна
            return ISO7816.FileStructure.DataObject;
        }}
        ///////////////////////////////////////////////////////////////////////////
        // прочитать содержимое файла
        ///////////////////////////////////////////////////////////////////////////
        public override Response ReadContent(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient)
        {
            // выполнить команду
            Response response = channel.SendCommand(secureType, 
                secureClient, INS.GetData, 0x00, 0x00, new byte[0], -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.Error(response)) return response; 
            
            // закодировать список тэгов
            byte[] encoded = DataCoding.Encode(new BER.TagList()); 
        
            // выполнить команду
            Response responseBERTLV = channel.SendCommand(secureType, 
                secureClient, INS.GetDataBERTLV, 0x00, 0x00, encoded, -1
            ); 
            // при наличии сокращенного идентификатора
            if (Response.Error(responseBERTLV) && ShortID.HasValue)
            {
                // выполнить команду
                Response responseShort = channel.SendCommand(secureType, 
                    secureClient, INS.GetDataBERTLV, 0x00, ShortID.Value, encoded, -1
                ); 
                // проверить отсутствие ошибок
                if (!Response.Error(responseShort)) responseBERTLV = responseShort; 
            }
            // проверить отсутствие ошибок
            return (!Response.Error(responseBERTLV)) ? responseBERTLV : response; 
        }
        // прочитать файл объектов
        public DataObject[] ReadBERTLVs(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient, bool interindustry)
        {
            // определить структуру файла
            FileStructure fileStructure = FileStructure; 
        
            // для файла записей и файла объектов
            if (fileStructure != ISO7816.FileStructure.DataObject && 
                fileStructure != ISO7816.FileStructure.DataObjectBERTLV)
            {
                // при ошибке выбросить исключение
                throw new ResponseException(0x6981); 
            }
            // прочитать содержимое файла
            Response response = ReadContent(channel, secureType, secureClient); 
        
            // проверить отсутствие ошибок
            ResponseException.Check(response);
        
            // раскодировать объекты
            return DataCoding.Decode(response.Data, interindustry); 
        }
        // прочитать файл объектов
        public SimpleTLV[] ReadSimpleTLVs(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient)
        {
            // определить структуру файла
            FileStructure fileStructure = FileStructure; 
        
            // для файла записей и файла объектов
            if (fileStructure != ISO7816.FileStructure.DataObject && 
                fileStructure != ISO7816.FileStructure.DataObjectSimpleTLV)
            {
                // при ошибке выбросить исключение
                throw new ResponseException(0x6981); 
            }
            // прочитать содержимое файла
            Response response = ReadContent(channel, secureType, secureClient); 
        
            // проверить отсутствие ошибок
            ResponseException.Check(response);
        
            // раскодировать объекты
            return SimpleTLV.Decode(response.Data); 
        }
        ///////////////////////////////////////////////////////////////////////////
        // прочитать объекты
        ///////////////////////////////////////////////////////////////////////////
        public SimpleTLV ReadObject(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient, int tag)
        {
            // проверить корректноть тэга
            if (tag < 0 || tag > 255) throw new ArgumentException(); 

            // выполнить команду
            Response response = channel.SendCommand(secureType, 
                secureClient, INS.GetData, 0x02, (byte)tag, new byte[0], -1
            ); 
            // проверить отсутствие ошибок
            ResponseException.Check(response);
        
            // проверить наличие данных
            if (response.Data.Length == 0) return null; 

            // раскодировать объект
            return SimpleTLV.Decode(response.Data)[0]; 
        }
        // прочитать объект
        public DataObject ReadObject(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient, Tag tag, bool interindustry) 
        {
            // закодировать тэг
            byte[] encoded = tag.Encoded; if (encoded.Length == 1)
            {
                // выполнить команду
                Response response = channel.SendCommand(secureType, 
                    secureClient, INS.GetData, 0x00, encoded[0], new byte[0], -1
                ); 
                // проверить отсутствие ошибок
                if (!Response.Error(response))
                {
                    // проверить наличие данных
                    if (response.Data.Length == 0) return null; 

                    // раскодировать объект
                    return DataCoding.Decode(encoded, interindustry)[0]; 
                }
            }
            else if (encoded.Length == 2)
            {
                // выполнить команду
                Response response = channel.SendCommand(secureType, 
                    secureClient, INS.GetData, encoded[0], encoded[1], new byte[0], -1
                ); 
                // проверить отсутствие ошибок
                if (!Response.Error(response))
                {
                    // проверить наличие данных
                    if (response.Data.Length == 0) return null; 

                    // раскодировать объект
                    return DataCoding.Decode(encoded, interindustry)[0]; 
                }
            }
            // прочитать объекты
            return ReadObjects(channel, secureType, 
                secureClient, new Tag[] { tag }, interindustry)[0];
        }
        // прочитать объекты
        public DataObject[] ReadObjects(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient, Tag[] tags, bool interindustry) 
        {
            // закодировать список тэгов
            byte[] encoded = DataCoding.Encode(new BER.TagList(tags)); 

            // выполнить команду
            Response response = channel.SendCommand(secureType, 
                secureClient, INS.GetDataBERTLV, 0x00, 0x00, encoded, -1
            ); 
            // при наличии сокращенного идентификатора
            if (Response.Error(response) && ShortID.HasValue)
            {
                // выполнить команду
                Response responseShort = channel.SendCommand(secureType, 
                    secureClient, INS.GetDataBERTLV, 0x00, ShortID.Value, encoded, -1
                ); 
                // проверить отсутствие ошибок
                if (!Response.Error(responseShort)) response = responseShort; 
            }
            // проверить отсутствие ошибок
            ResponseException.Check(response); 
            
            // раскодировать объекты
            return DataCoding.Decode(response.Data, interindustry); 
        }
        // прочитать объекты
        public DataObject[] ReadObjects(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient, Header[] headers, bool interindustry)
        {
            // закодировать список заголовков
            byte[] encoded = DataCoding.Encode(new BER.HeaderList(headers)); 

            // выполнить команду
            Response response = channel.SendCommand(secureType, 
                secureClient, INS.GetDataBERTLV, 0x00, 0x00, encoded, -1
            ); 
            // при наличии сокращенного идентификатора
            if (Response.Error(response) && ShortID.HasValue)
            {
                // выполнить команду
                Response responseShort = channel.SendCommand(secureType, 
                    secureClient, INS.GetDataBERTLV, 0x00, ShortID.Value, encoded, -1
                ); 
                // проверить отсутствие ошибок
                if (!Response.Error(responseShort)) response = responseShort; 
            }
            // проверить отсутствие ошибок
            ResponseException.Check(response); 
            
            // раскодировать объекты
            return DataCoding.Decode(response.Data, interindustry); 
        }
        // прочитать объекты
        public DataObject[] ReadObjects(LogicalChannel channel, 
            SecureType secureType, SecureClient secureClient, 
            ExtendedHeader[] extendedHeaders, bool interindustry)
        {
            // закодировать список заголовков
            byte[] encoded = DataCoding.Encode(new BER.ExtendedHeaderList(extendedHeaders)); 

            // выполнить команду
            Response response = channel.SendCommand(secureType, 
                secureClient, INS.GetDataBERTLV, 0x00, 0x00, encoded, -1
            ); 
            // при наличии сокращенного идентификатора
            if (Response.Error(response) && ShortID.HasValue)
            {
                // выполнить команду
                Response responseShort = channel.SendCommand(secureType, 
                    secureClient, INS.GetDataBERTLV, 0x00, ShortID.Value, encoded, -1
                ); 
                // проверить отсутствие ошибок
                if (!Response.Error(responseShort)) response = responseShort; 
            }
            // проверить отсутствие ошибок
            ResponseException.Check(response); 
            
            // раскодировать объекты
            return DataCoding.Decode(response.Data, interindustry); 
        }
    }
}

File: Source/Aladdin.ISO7816.Java/src/aladdin/iso7816/DataObjectFile.java
package aladdin.iso7816;
import aladdin.iso7816.ber.*; 
import aladdin.asn1.*;
import java.io.*;

///////////////////////////////////////////////////////////////////////////////
// Элементарный файл объектов
///////////////////////////////////////////////////////////////////////////////
public class DataObjectFile extends ElementaryFile
{
    // конструктор
    protected DataObjectFile(DedicatedFile parent, short id) throws IOException
    { 
        // сохранить переданные параметры
        super(parent, id); 
    }
    // конструктор
    protected DataObjectFile(DedicatedFile parent, byte shortID) throws IOException
    { 
        // сохранить переданные параметры
        super(parent, shortID);
    }
    // конструктор
    protected DataObjectFile(DedicatedFile parent, 
        Short id, Byte shortID, FileControlInformation info) throws IOException
    { 
        // сохранить переданные параметры
        super(parent, id, shortID, info); 
    }
    // структура файла
    @Override public FileStructure fileStructure()
    {
        // получить дескриптор файла
        DataObject[] objs = info().get(Tag.context(0x02, PC.PRIMITIVE)); 
            
        // проверить наличие дескриптора
        if (objs.length == 0) return FileStructure.DATA_OBJECT; 

        // получить содержимое
        byte[] content = objs[0].content(); 
        
        // проверить размер содержимого
        if (content.length < 1 || (content[0] & 0x80) != 0)
        {
            // указать значение по умолчанию
            return FileStructure.DATA_OBJECT; 
        }
        // в зависимости установленных битов
        if (((content[0] >>> 3) & 0x7) == 0x7)
        {
            // в зависимости установленных битов
            switch (content[0] & 0x7)
            {
            case 0x1: return FileStructure.DATA_OBJECT_BERTLV;
            case 0x2: return FileStructure.DATA_OBJECT_SIMPLETLV;
            }
        }
        // структура файла неизвестна
        return FileStructure.DATA_OBJECT;
    }
    // выделить родительский каталог
    @Override public DedicatedFile selectParent(LogicalChannel channel) throws IOException
    {
        // выделить родительский каталог
        parent().selectFromChild(channel); return parent(); 
    }
    ///////////////////////////////////////////////////////////////////////////
    // прочитать содержимое файла
    ///////////////////////////////////////////////////////////////////////////
    @Override public Response readContent(LogicalChannel channel, 
        int secureType, SecureClient secureClient) throws IOException
    {
        // выполнить команду
        Response response = channel.sendCommand(secureType, 
            secureClient, INS.GET_DATA, (byte)0x00, (byte)0x00, new byte[0], -1
        ); 
        // проверить отсутствие ошибок
        if (!Response.error(response)) return response; 
            
        // закодировать список тэгов
        byte[] encoded = dataCoding().encode(new TagList()); 
        
        // выполнить команду
        Response responseBERTLV = channel.sendCommand(secureType, 
            secureClient, INS.GET_DATA_BERTLV, (byte)0x00, (byte)0x00, encoded, -1
        ); 
        // при наличии сокращенного идентификатора
        if (Response.error(responseBERTLV) && shortID() != null)
        {
            // выполнить команду
            Response responseShort = channel.sendCommand(secureType, 
                secureClient, INS.GET_DATA_BERTLV, (byte)0x00, shortID(), encoded, -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.error(responseShort)) responseBERTLV = responseShort; 
        }
        // проверить отсутствие ошибок
        return (!Response.error(responseBERTLV)) ? responseBERTLV : response; 
    }
    // прочитать файл объектов
    public DataObject[] readBERTLVs(LogicalChannel channel, 
        int secureType, SecureClient secureClient, boolean interindustry) throws IOException
    {
        // определить структуру файла
        FileStructure fileStructure = fileStructure(); 
        
        // для файла записей и файла объектов
        if (fileStructure != FileStructure.DATA_OBJECT && 
            fileStructure != FileStructure.DATA_OBJECT_BERTLV)
        {
            // при ошибке выбросить исключение
            throw new ResponseException((short)0x6981); 
        }
        // прочитать содержимое файла
        Response response = readContent(channel, secureType, secureClient); 
        
        // проверить отсутствие ошибок
        ResponseException.check(response);
        
        // раскодировать объекты
        return dataCoding().decode(response.data, interindustry); 
    }
    // прочитать файл объектов
    public SimpleTLV[] readSimpleTLVs(LogicalChannel channel, 
        int secureType, SecureClient secureClient) throws IOException
    {
        // определить структуру файла
        FileStructure fileStructure = fileStructure(); 
        
        // для файла записей и файла объектов
        if (fileStructure != FileStructure.DATA_OBJECT && 
            fileStructure != FileStructure.DATA_OBJECT_SIMPLETLV)
        {
            // при ошибке выбросить исключение
            throw new ResponseException((short)0x6981); 
        }
        // прочитать содержимое файла
        Response response = readContent(channel, secureType, secureClient); 
        
        // проверить отсутствие ошибок
        ResponseException.check(response);
        
        // раскодировать объекты
        return SimpleTLV.decode(response.data); 
    }
    ///////////////////////////////////////////////////////////////////////////
    // прочитать объекты
    ///////////////////////////////////////////////////////////////////////////
    public SimpleTLV readObject(LogicalChannel channel, 
        int secureType, SecureClient secureClient, int tag) throws IOException
    {
        // проверить корректноть тэга
        if (tag < 0 || tag > 255) throw new IllegalArgumentException(); 

        // выполнить команду
        Response response = channel.sendCommand(secureType, 
            secureClient, INS.GET_DATA, (byte)0x02, (byte)tag, new byte[0], -1
        ); 
        // проверить отсутствие ошибок
        ResponseException.check(response);
        
        // проверить наличие данных
        if (response.data.length == 0) return null; 

        // раскодировать объект
        return SimpleTLV.decode(response.data)[0]; 
    }
    // прочитать объект
    public DataObject readObject(LogicalChannel channel, 
        int secureType, SecureClient secureClient, 
        Tag tag, boolean interindustry) throws IOException
    {
        // закодировать тэг
        byte[] encoded = tag.encoded; if (encoded.length == 1)
        {
            // выполнить команду
            Response response = channel.sendCommand(secureType, 
                secureClient, INS.GET_DATA, (byte)0x00, encoded[0], new byte[0], -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.error(response))
            {
                // проверить наличие данных
                if (response.data.length == 0) return null; 

                // раскодировать объект
                return dataCoding().decode(encoded, interindustry)[0]; 
            }
        }
        else if (encoded.length == 2)
        {
            // выполнить команду
            Response response = channel.sendCommand(secureType, 
                secureClient, INS.GET_DATA, encoded[0], encoded[1], new byte[0], -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.error(response))
            {
                // проверить наличие данных
                if (response.data.length == 0) return null; 

                // раскодировать объект
                return dataCoding().decode(encoded, interindustry)[0]; 
            }
        }
        // прочитать объекты
        return readObjects(channel, secureType, 
            secureClient, new Tag[] { tag }, interindustry)[0];
    }
    // прочитать объекты
    public DataObject[] readObjects(LogicalChannel channel, 
        int secureType, SecureClient secureClient, 
        Tag[] tags, boolean interindustry) throws IOException
    {
        // закодировать список тэгов
        byte[] encoded = dataCoding().encode(new TagList(tags)); 

        // выполнить команду
        Response response = channel.sendCommand(secureType, 
            secureClient, INS.GET_DATA_BERTLV, (byte)0x00, (byte)0x00, encoded, -1
        ); 
        // при наличии сокращенного идентификатора
        if (Response.error(response) && shortID() != null)
        {
            // выполнить команду
            Response responseShort = channel.sendCommand(secureType, 
                secureClient, INS.GET_DATA_BERTLV, (byte)0x00, shortID(), encoded, -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.error(responseShort)) response = responseShort; 
        }
        // проверить отсутствие ошибок
        ResponseException.check(response); 
        
        // раскодировать объекты
        return dataCoding().decode(response.data, interindustry); 
    }
    // прочитать объекты
    public DataObject[] readObjects(LogicalChannel channel, 
        int secureType, SecureClient secureClient, 
        Header[] headers, boolean interindustry) throws IOException
    {
        // закодировать список заголовков
        byte[] encoded = dataCoding().encode(new HeaderList(headers)); 

        // выполнить команду
        Response response = channel.sendCommand(secureType, 
            secureClient, INS.GET_DATA_BERTLV, (byte)0x00, (byte)0x00, encoded, -1
        ); 
        // при наличии сокращенного идентификатора
        if (Response.error(response) && shortID() != null)
        {
            // выполнить команду
            Response responseShort = channel.sendCommand(secureType, 
                secureClient, INS.GET_DATA_BERTLV, (byte)0x00, shortID(), encoded, -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.error(responseShort)) response = responseShort; 
        }
        // проверить отсутствие ошибок
        ResponseException.check(response); 
        
        // раскодировать объекты
        return dataCoding().decode(response.data, interindustry); 
    }
    // прочитать объекты
    public DataObject[] readObjects(LogicalChannel channel, 
        int secureType, SecureClient secureClient, 
        ExtendedHeader[] extendedHeaders, boolean interindustry) throws IOException 
    {
        // закодировать список заголовков
        byte[] encoded = dataCoding().encode(new ExtendedHeaderList(extendedHeaders)); 

        // выполнить команду
        Response response = channel.sendCommand(secureType, 
            secureClient, INS.GET_DATA_BERTLV, (byte)0x00, (byte)0x00, encoded, -1
        ); 
        // при наличии сокращенного идентификатора
        if (Response.error(response) && shortID() != null)
        {
            // выполнить команду
            Response responseShort = channel.sendCommand(secureType, 
                secureClient, INS.GET_DATA_BERTLV, (byte)0x00, shortID(), encoded, -1
            ); 
            // проверить отсутствие ошибок
            if (!Response.error(responseShort)) response = responseShort; 
        }
        // проверить отсутствие ошибок
        ResponseException.check(response); 
        
        // раскодировать объекты
        return dataCoding().decode(response.data, interindustry); 
    }
}
